File: transpiler/javatests/com/google/j2cl/integration/java/arithmeticexception/Main.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package arithmeticexception;

import static com.google.j2cl.integration.testing.Asserts.assertThrowsArithmeticException;

public class Main {
  public static void main(String... args) {
    testDivideByZero();
    testModByZero();
  }

  private static void testDivideByZero() {
    assertThrowsArithmeticException(
        () -> {
          int a = 10;
          int b = 0;
          int unused = a / b;
        });

    assertThrowsArithmeticException(
        () -> {
          int unused = 10;
          int b = 0;
          unused /= b;
        });

    assertThrowsArithmeticException(
        () -> {
          int a = 10;
          int b = 0;
          int unused = 1 + a / b;
        });

    assertThrowsArithmeticException(
        () -> {
          short a = 10;
          short b = 0;
          short unused = (short) (a / b);
        });

    assertThrowsArithmeticException(
        () -> {
          char a = 10;
          char b = 0;
          char unused = (char) (a / b);
        });

    assertThrowsArithmeticException(
        () -> {
          byte a = 10;
          byte b = 0;
          @SuppressWarnings("unused")
          byte unused = (byte) (a / b);
        });

    assertThrowsArithmeticException(
        () -> {
          long a = 10;
          long b = 0;
          long unused = (a / b);
        });

    assertThrowsArithmeticException(
        () -> {
          int a = 10;
          int b = 0;
          long unused = 10L + (a / b);
        });

    assertThrowsArithmeticException(
        () -> {
          short a = 10;
          short b = 0;
          long unused = 10L + (a / b);
        });
  }

  private static void testModByZero() {
    assertThrowsArithmeticException(
        () -> {
          int a = 10;
          int b = 0;
          int unused = a % b;
        });

    assertThrowsArithmeticException(
        () -> {
          int unused = 10;
          int b = 0;
          unused %= b;
        });

    assertThrowsArithmeticException(
        () -> {
          short a = 10;
          short b = 0;
          short unused = (short) (a % b);
        });

    assertThrowsArithmeticException(
        () -> {
          char a = 10;
          char b = 0;
          char unused = (char) (a % b);
        });

    assertThrowsArithmeticException(
        () -> {
          byte a = 10;
          byte b = 0;
          byte unused = (byte) (a % b);
        });

    assertThrowsArithmeticException(
        () -> {
          long a = 10;
          long b = 0;
          long unused = (a % b);
        });

    assertThrowsArithmeticException(
        () -> {
          int a = 10;
          int b = 0;
          long unused = 10L + (a % b);
        });

    assertThrowsArithmeticException(
        () -> {
          short a = 10;
          short b = 0;
          double unused = 10d + (a % b);
        });
  }
}

File: transpiler/javatests/com/google/j2cl/integration/kotlin/arithmeticexception/Main.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package arithmeticexception

import com.google.j2cl.integration.testing.Asserts.assertThrowsArithmeticException

fun main(vararg unused: String) {
  testDivideByZero()
  testModByZero()
}

private fun testDivideByZero() {
  assertThrowsArithmeticException {
    val a = 10
    val b = 0
    val unused = a / b
  }

  assertThrowsArithmeticException {
    var unused = 10
    val b = 0
    unused /= b
  }

  assertThrowsArithmeticException {
    val a = 10
    val b = 0
    val unused = 1 + a / b
  }

  assertThrowsArithmeticException {
    val a: Short = 10
    val b: Short = 0
    val unused = (a / b).toShort()
  }

  // No division operation available on Char
  // assertThrowsArithmeticException {
  //   val a = 10.toChar()
  //   val b = 0.toChar()
  //   val unused: Char = (a / b).toChar()
  // }
  assertThrowsArithmeticException {
    val a: Byte = 10
    val b: Byte = 0
    val unused = (a / b).toByte()
  }

  assertThrowsArithmeticException {
    val a: Long = 10
    val b: Long = 0
    val unused = (a / b)
  }

  assertThrowsArithmeticException {
    val a = 10
    val b = 0
    val unused = 10L + (a / b)
  }

  assertThrowsArithmeticException {
    val a: Short = 10
    val b: Short = 0
    val unused = 10L + (a / b)
  }
}

private fun testModByZero() {
  assertThrowsArithmeticException {
    val a = 10
    val b = 0
    val unused = a % b
  }

  assertThrowsArithmeticException {
    var unused = 10
    val b = 0
    unused %= b
  }

  assertThrowsArithmeticException {
    val a: Short = 10
    val b: Short = 0
    val unused = (a % b).toShort()
  }

  // No modulo operation available on Char
  // assertThrowsArithmeticException {
  //   val a = 10.toChar()
  //   val b = 0.toChar()
  //   val unused: Char = (a % b).toChar()
  // }

  assertThrowsArithmeticException {
    val a: Byte = 10
    val b: Byte = 0
    val unused = (a % b).toByte()
  }

  assertThrowsArithmeticException {
    val a: Long = 10
    val b: Long = 0
    val unused = (a % b)
  }

  assertThrowsArithmeticException {
    val a = 10
    val b = 0
    val unused = 10L + (a % b)
  }

  assertThrowsArithmeticException {
    val a: Short = 10
    val b: Short = 0
    val unused = 10.0 + (a % b)
  }
}
